pq-crypto/EVP_kem.c

```diff
@@ -0,0 +1,163 @@
+/********************************************************************************************
+* KEM Library
+*
+* Abstract: Post-Quantum Algorithm API
+*********************************************************************************************/
+
+#include <stdint.h>
+#include <stdlib.h>
+#include "EVP_kem.h"
+#include "EVP_pq_constants.h"
+
+struct pq_kem evp_sike_p434_r3 = {
+        .name = "SIKEp434r3-KEM",
+        .public_key_length = SIKE_P434_R3_PUBLIC_KEY_BYTES,
+        .private_key_length = SIKE_P434_R3_PRIVATE_KEY_BYTES,
+        .shared_secret_key_length = SIKE_P434_R3_SHARED_SECRET_BYTES,
+        .ciphertext_length = SIKE_P434_R3_CIPHERTEXT_BYTES,
+        .generate_keypair = &evp_sike_p434_r3_crypto_kem_keypair,
+        .encapsulate = &evp_sike_p434_r3_crypto_kem_enc,
+        .decapsulate = &evp_sike_p434_r3_crypto_kem_dec,
+};
+
+const struct pq_kem_params evp_sike_kem_params = {
+    .kem = &evp_sike_p434_r3,
+    .public_key = 0,
+    .private_key = 0,
+    .ciphertext = 0,
+    .shared_secret = 0,
+};
+
+int pq_kem_params_alloc(pq_kem_params *key_params) {
+    key_params->public_key = malloc(key_params->kem->public_key_length);
+    key_params->private_key = malloc(key_params->kem->private_key_length);
+    key_params->ciphertext = malloc(key_params->kem->ciphertext_length);
+    key_params->shared_secret = malloc(key_params->kem->shared_secret_key_length);
+
+    return EVP_PQ_SUCCESS;
+}
+
+int pq_kem_params_free(pq_kem_params *key_params) {
+    if (key_params != NULL) {
+        free(key_params->public_key);
+        free(key_params->private_key);
+        free(key_params->ciphertext);
+        free(key_params->shared_secret);
+        return EVP_PQ_SUCCESS;
+    }
+
+    return EVP_PQ_FAILURE;
```

Comment:
Is this really failure? We definitely want to verify `key_params != NULL` above, but if key_params is NULL then I'm not sure that means that the function should return failure. It's kind of like "we freed everything that you told us to, but all you gave us was a NULL pointer so there was nothing to free, so we did our job!" which I think counts as success.

Reply:
I see that makes sense. I was thinking about saving the user from passing in null pointers but looking at it now it would be a completely different functionality.